dev-containers-case-study (#178)
NoroffFEU/development-platforms: src/content/docs/devtools/Dev-Containers/dev-containers-case-study.md

---
title: VSCode Dev Containers case study
author: Ingeborg Sanna / Spekkhogger
tags: devcontainers, containers, docker, VSCode, microsoft
---

## Introduction
Dev Containers is a feature within Visual Studio Code, and is used in combination with Docker. It is a tool that streamlines the process of setting up and managing a development enviorment. It enables the developers to save all tools, dependencies, and configurations within a container, making it easy to share and maintain development enviornments across machines and team members.

> Combining Dev Containers with Docker enhances the development experience by simplifying environment setup, ensuring consistency, and providing a smooth workflow within Visual Studio Code. While Docker is a powerful containerization tool, Dev Containers make it more accessible and developer-friendly, particularly for those who prefer using Visual Studio Code

Reviewer: ```suggestion
Combining Dev Containers with Docker enhances the development experience by simplifying environment setup, ensuring consistency, and providing a smooth workflow within Visual Studio Code. While Docker is a powerful containerization tool, Dev Containers make it more accessible and developer-friendly, particularly for those who prefer using Visual Studio Code.
```

## Brief History

> As Dev Containers is a relatively new feature it's history is truly brief. The history of Dev Containers lay within the history of Remote Development, and was only recently separated and made into it's own extension. 

Reviewer: ```suggestion
As Dev Containers is a relatively new feature it's history is truly brief. The history of Dev Containers lay within the history of Remote Development, and was only recently separated and made into its own extension. 
```

- 2022 - Present: Microsoft and the open source community continue to improve features.

## Main Features
| Feature:          | Description:                                                                                                                                                                       |
|---|---|
| Reproducibility   | Dev Containers ensure that all team members have identical development environments. This reduces the "it works on my machine" problem and minimizes configuration-related issues. |
| Consistency       | Dev Containers standardize the development process, ensuring that the same tools and configurations are used.                                                                      |
| Scalability       | Containers can be easily scaled and managed, making it convenient for projects with complex dependencies or multiple microservices.                                                |
| Security          | Dev Containers provide a layer of security by isolating development environments. Any changes made within the container do not affect the host system.                             |

> | Templates(Images) | Templates provide preconfigured settings and dependencies tailored to a specific languages.     

Reviewer: ```suggestion
| Templates(Images) | Templates provide preconfigured settings and dependencies tailored to specific languages.     
```

> #### Notaable weaknesses

Reviewer: ```suggestion
## Notable weaknesses
```

1. Dev Containers ties you to use both VSCode and Docker. It is currently being made available to other code editors, but the 1st class integration it has towards VSCode is so far unmatched. 
2. The setup and learning curve of Dev Containers can be complex, especially for those without prior experience with Docker or similar containers. 

## Market Comparison
#### A short comparison between Dev Containers and Codespaces(Github): 
| Comparison | Description: |
|---|---|
| Use cases | Dev Containers are ideal for local development and those who prefer Visual Studio Code as their code editor. Codespaces are well-suited for collaborative development, remote work, and projects hosted on GitHub. |
| Location | Dev Containers are local and run on your machine, while Codespaces are hosted remotely in the cloud. |
| Integration | Dev Containers are integrated with Visual Studio Code, while Codespaces are closely integrated with GitHub, making them an attractive choice for projects hosted on GitHub. |
| Customization | Dev Containers provide more options in customizing development environments. Codespaces might have some restrictions. |

> In summary both provide you with great tools that streamlines the process of setting up and managing development enviorments. Which you should choose will depend on the needs of your project. While Dev Containers are primarily used for local development, Codespaces is a remote and collaborative enviorment. They both tie you to other tools. With Dev Containers you will be tied to VSCode and Docker, and Codespaces with tie you to Github. For a more custamizable enviornment, Dev Containers will be the better choice. 

Reviewer: ```suggestion
In summary, both provide you with great tools that streamline the process of setting up and managing development environments. Which you should choose will depend on the needs of your project. While Dev Containers are primarily used for local development, Codespaces is a remote and collaborative environment. They both tie you to other tools. With Dev Containers you will be tied to VSCode and Docker, and Codespaces will tie you to Github. For a more customizable environment, Dev Containers will be the better choice. 
```

Note: Both are owned by Microsoft. 

## Getting Started

> This guide will show you how to open an existing project using one of templates provided by Dev Containers. 

Reviewer: ```suggestion
This guide will show you how to open an existing project using one of the templates provided by Dev Containers. 
```
